### DISCUSSION TITLE
Automatic snippet expansion and specific snippets in math mode

### DISCUSSION THREAD

**Discussion by TryerGit** [James-Yu/LaTeX-Workshop#3615] [Q&A] [answered]
Hello,

I have been using Vim with VimTeX  and Ultisnips as my LaTeX front end and I would now like to try VSCode-LaTeX-Workshop out.

I would like to check if the following features of Vim + VimTeX + Ultisnips are available within VSCode-LaTeX-Workshop.

(I use VSCode for C++ programming and am aware that atleast within C++, not all of the following are possible)

(1) Detecting math environment:

```
context "math()"
snippet ss "subscript" iA
_{$1} $0
endsnippet
```

Within `$ | $` or `\( | \)` or `\[  | \]`, where ` | ` is the cursor position, typing `ss` automatically gives the subscript command. The key here is that this is *automatic* because the option `iA` [this comes from Ultisnips] expands the snippet automatically without having to press a snippet trigger/expansion/completion (for e.g., Tab) key. The `context "math()"` line ensures that this expansion occurs only in math delimited environments and not plain text prose environments. I have looked at https://github.com/James-Yu/LaTeX-Workshop/wiki/Snippets#Mathematical-font-commands and it appears to me that pressing the snippet trigger/completion/expansion key (say the Tab key) is mandatory. Is my understanding correct?

(2) Somewhat related to the above, is it possible to have the equivalent of the following snippet:

```
snippet mm "Inline Math" wA
\\( $1 \\) $0
endsnippet

```

This snippet has option `wA` and the functionality here is that if the user types `mm` (the `wA` option stands for the beginning of a new word and the expansion is automatic with no need to press a snippet trigger/expansion/completion (e.g., Tab) key.  Within a line, this provides the `\( | \)` inline math environment. I think this is not possible at least within C++ programming in VSCode -- every snippet expansion seems to require the snippet trigger/expansion/completion key. I would like to confirm that it is indeed possible/not possible within LaTeX-Workshop as well.

The use case is that pressing `mm` is a lot easier since the keys are close to the home row rather than pressing any other snippet trigger/expansion/completion key

Thank you.

----

I have done a search for `automatic expansion` in the issues search tab and that did not turn any relevant issues/discussions. Hence this OP.

**Comment by James-Yu** [comment, marked as answer]
No, it's not possible with vscode, not limited by the extension. Math mode detection is theoretically possible but unlikely implemented, due to the extra computing burden.

**Reply by TryerGit** [reply]
Fair enough. Thank you.

**Comment by TryerGit** [comment]
For what it is worth, I have opened an issue with the VSCode folks requesting this feature:

https://github.com/microsoft/vscode/issues/170423

**Reply by TryerGit** [reply]
20 upvotes needed on the above for any hope to get it ~~done~~ started in the hope that it completes in 2-3 years. Pretty please :-)

**Comment by sleepymalc** [comment]
Nah, it's already implemented, but it just needed to be more well-known. Please check out [HyperSnips](https://marketplace.visualstudio.com/items?itemName=draivin.hsnips). I also have documentation on how to set it up correctly and have some showcases [here](https://github.com/sleepymalc/VSCode-LaTeX-Inkscape); for the exact section, see [here](https://github.com/sleepymalc/VSCode-LaTeX-Inkscape#hypersnips).

**Reply by TryerGit** [reply]
@sleepymalc 

Wow, that is good to know. I will definitely give it a shot. I have to state that basic Vim + Ultisnips is extraordinarily fast and in my experience, VSCode snippets tend to be a bit laggy/slow and do not tend to trigger unless Ctrl-Space is used. Looking forward to trying out your suggestion.

**Reply by sleepymalc** [reply]
Yea, Hypersnips definitely can have some improvement, but some of the recent updates really make it stable (tho it's fast throughout, in fact).

**Reply by TryerGit** [reply]
@sleepymalc  I have begun trying HyperSnips and so far they seem to be doing quite fine and fast -- indeed they seem faster than native VSCode snippet triggering. I am curious to know how you handle the context "math()" snippets. For e.g., the following was one of my snippets in Vim + VimTeX + Ultisnips:

```
context "math()"
snippet ss "subscript" iA
_{$1} $0
endsnippet
```
that triggers automatically but only in math mode. 

Within only a math environment (display math or inline math environment) but not within prose, how would it be possible to have the above snippet triggered using Hypersnips? As I recall, the context math() code was provided by VimTeX and not by UltiSnips.

Thanks.

**Reply by sleepymalc** [reply]
It's at the top of this [file](https://github.com/sleepymalc/VSCode-LaTeX-Inkscape/blob/main/VSCode-setting/Snippets/latex.hsnips). I guess it's the best way to handle math context since sometimes you might run into cases that you need to use `\text{}` inside of a math environment, or even use math environment again in a nested manner.

**Reply by TryerGit** [reply]
Yes, I tried that and it works perfectly. I have to say that Hypersnips is severly underrated. I can imagine using that for many of my C++ snippet needs as well. I am curious to know -- given Hypersnips, do you have any need for native VSCode snippets at all -- either in LaTeX or C++ in case you program in it.

**Reply by sleepymalc** [reply]
Yes, in my set-up, when I need to insert a snippet via VSCode in a sequence of tasks, VSCode snippets will be helpful. But apart form that, I only use Hypersnips.
